File: frontend/src/api/modules/disks.ts
/**
 * @file disks.ts
 * @description 节点磁盘管理稳定领域 API。
 */

import type { AxiosRequestConfig } from 'axios'
import http from '@/api'
import type {
  CreateDiskOperationRequest,
  DiskDetail,
  DiskInventory,
  DiskOperation,
} from '@/api/generated/disks'

const config = (signal?: AbortSignal): AxiosRequestConfig => ({ signal })
const nodePath = (nodeId: string) => `/node/${encodeURIComponent(nodeId)}`

export const disksApi = {
  inventory: (nodeId: string, signal?: AbortSignal) =>
    http.get<DiskInventory>(`${nodePath(nodeId)}/disks`, undefined, config(signal)),
  detail: (nodeId: string, diskId: string, signal?: AbortSignal) =>
    http.get<DiskDetail>(
      `${nodePath(nodeId)}/disks/${encodeURIComponent(diskId)}`,
      undefined,
      config(signal),
    ),
  createOperation: (nodeId: string, request: CreateDiskOperationRequest, idempotencyKey: string) =>
    http.post<DiskOperation>(`${nodePath(nodeId)}/disk-operations`, request, {
      headers: { 'Idempotency-Key': idempotencyKey },
    }),
  operation: (nodeId: string, operationId: string, signal?: AbortSignal) =>
    http.get<DiskOperation>(
      `${nodePath(nodeId)}/disk-operations/${encodeURIComponent(operationId)}`,
      undefined,
      config(signal),
    ),
  cancelOperation: (nodeId: string, operationId: string) =>
    http.post<DiskOperation>(
      `${nodePath(nodeId)}/disk-operations/${encodeURIComponent(operationId)}/cancel`,
    ),
}

export type {
  CreateDiskOperationRequest,
  DiskDetail,
  DiskInventory,
  DiskOperation,
  DiskPartition,
  DiskSummary,
} from '@/api/generated/disks'

File: crates/seclab-agent/src/api/disks.rs
//! Agent 磁盘内部 API：提供可信事实清单和按需详情。

use crate::services::{disk_inventory, disk_operations};
use crate::state::AppState;
use crate::types::{ApiResponse, ApiResult};
use axum::{
    Json, Router,
    extract::{FromRequestParts, Path, State},
    http::{HeaderName, StatusCode},
    response::{IntoResponse, Response},
    routing::{get, post},
};
use seclab_contracts::{api::ErrorCode, disks::AgentCreateDiskOperationRequest};
use std::sync::Arc;

const ACTOR_KIND_HEADER: HeaderName = HeaderName::from_static("x-seclab-actor-kind");
const ACTOR_NAME_HEADER: HeaderName = HeaderName::from_static("x-seclab-actor-name");

/// 仅接受 Master 内部链路注入的可信操作上下文。
struct TrustedOperationContext;

impl<S> FromRequestParts<S> for TrustedOperationContext
where
    S: Send + Sync,
{
    type Rejection = crate::types::ApiError;

    async fn from_request_parts(
        parts: &mut axum::http::request::Parts,
        _state: &S,
    ) -> Result<Self, Self::Rejection> {
        let kind = parts
            .headers
            .get(&ACTOR_KIND_HEADER)
            .and_then(|value| value.to_str().ok());
        let name = parts
            .headers
            .get(&ACTOR_NAME_HEADER)
            .and_then(|value| value.to_str().ok())
            .map(str::trim)
            .filter(|value| !value.is_empty());
        if matches!(kind, Some("user" | "system")) && name.is_some() {
            Ok(Self)
        } else {
            Err(crate::types::ApiError::forbidden(
                ErrorCode::AuthForbidden,
                "trusted operation context is required",
            ))
        }
    }
}

/// 返回节点磁盘摘要。
async fn inventory(State(state): State<Arc<AppState>>) -> ApiResult<Response> {
    let value = disk_inventory::inventory(&state.metadata_db).await?;
    Ok(ApiResponse::success_with_raw("Disk inventory loaded", value).into_response())
}

/// 返回单块磁盘详情。
async fn detail(
    State(state): State<Arc<AppState>>,
    Path(disk_id): Path<String>,
) -> ApiResult<Response> {
    let value = disk_inventory::detail(&state.metadata_db, &disk_id).await?;
    Ok(ApiResponse::success_with_raw("Disk detail loaded", value).into_response())
}

/// 幂等创建磁盘后台操作。
async fn create_operation(
    State(state): State<Arc<AppState>>,
    _trusted: TrustedOperationContext,
    Json(request): Json<AgentCreateDiskOperationRequest>,
) -> ApiResult<Response> {
    let value = disk_operations::submit(state, request).await?;
    Ok((
        StatusCode::ACCEPTED,
        Json(ApiResponse::success_with_raw(
            "Disk operation accepted",
            value,
        )),
    )
        .into_response())
}

/// 返回磁盘后台操作状态。
async fn operation(
    State(state): State<Arc<AppState>>,
    Path(operation_id): Path<String>,
) -> ApiResult<Response> {
    let value = disk_operations::get(&state, &operation_id).await?;
    Ok(ApiResponse::success_with_raw("Disk operation loaded", value).into_response())
}

/// 请求取消仍处于安全阶段的操作。
async fn cancel_operation(
    State(state): State<Arc<AppState>>,
    _trusted: TrustedOperationContext,
    Path(operation_id): Path<String>,
) -> ApiResult<Response> {
    let value = disk_operations::cancel(&state, &operation_id).await?;
    Ok((
        StatusCode::ACCEPTED,
        Json(ApiResponse::success_with_raw(
            "Disk operation cancellation accepted",
            value,
        )),
    )
        .into_response())
}

/// 构建磁盘内部路由。
pub fn disk_router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(inventory))
        .route("/operations", post(create_operation))
        .route("/operations/{operation_id}", get(operation))
        .route("/operations/{operation_id}/cancel", post(cancel_operation))
        .route("/{disk_id}", get(detail))
}
